stripecardscan/src/main/java/com/stripe/android/stripecardscan/scanui/ScanFragment.kt

```diff
@@ -0,0 +1,273 @@
+import com.stripe.android.camera.framework.Stats
+import com.stripe.android.stripecardscan.framework.StorageFactory
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.runBlocking
+import kotlin.coroutines.CoroutineContext
+
+private const val PERMISSION_RATIONALE_SHOWN = "permission_rationale_shown"
+
+abstract class ScanFragment : Fragment(), CoroutineScope {
+
+    override val coroutineContext: CoroutineContext = Dispatchers.Main
+
+    internal val scanStat = Stats.trackTask("scan_activity")
+    private val permissionStat = Stats.trackTask("camera_permission")
+
+    protected var isFlashlightOn: Boolean = false
+        private set
+
+    internal val cameraAdapter by lazy { buildCameraAdapter() }
+    private val cameraErrorListener by lazy {
+        CameraErrorListenerImpl(requireActivity()) { t -> scanFailure(t) }
+    }
+
+    /**
+     * The listener which will handle the results from the scan.
+     */
+    internal abstract val resultListener: ScanResultListener
+
+    private val storage by lazy {
+        StorageFactory.getStorageInstance(requireActivity(), "scan_camera_permissions")
+    }
+
+    private val cameraPermissionLauncher =
+        registerForActivityResult(
+            ActivityResultContracts.RequestPermission()
+        ) { isGranted: Boolean ->
+            if (isGranted) {
+                launch { permissionStat.trackResult("success") }
+                prepareCamera { onCameraReady() }
+            } else {
+                launch { permissionStat.trackResult("failure") }
+                showPermissionDenied()
+            }
+        }
+
+    override fun onStart() {
+        super.onStart()
+        Stats.startScan()
+
+        if (!CameraAdapter.isCameraSupported(requireActivity())) {
+            showCameraNotSupported()
+        }
+    }
+
+    override fun onResume() {
+        super.onResume()
+        hideSystemUi()
+
+        launch {
+            delay(1500)
+            hideSystemUi()
+        }
+
+        if (!cameraAdapter.isBoundToLifecycle()) {
+            ensurePermissionAndStartCamera()
+        }
+    }
+
+    protected open fun hideSystemUi() {
+        // Prevent screenshots and keep the screen on while scanning.
+        requireActivity().window.setFlags(
+            WindowManager.LayoutParams.FLAG_SECURE +
+                WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON,
+            WindowManager.LayoutParams.FLAG_SECURE +
+                WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON,
+        )
+    }
+
+    override fun onPause() {
+        super.onPause()
+        setFlashlightState(false)
+    }
+
+    /**
+     * Ensure that the camera permission is available. If so, start the camera. If not, request it.
+     */
+    protected open fun ensurePermissionAndStartCamera() {
```

Comment:
Can we share some logic between this fragment and the `CameraPermissionCheckingActivity`?

Reply:
I'm not sure how to do that since they're integrated with lifecycle callbacks

Reply:
you're right. we may want to eventually condense this into a `CameraPermissionCheckingFragment` in case we wind up building a cardverify fragment, but for this PR this is fine.